File: C++/source/green.cpp
#include "green.hpp"

// Constructor
GreensFunction::GreensFunction(mat V, float height, float width, float efm) {
    nc = V.n_cols;
    nr = V.n_rows;
    
    Potential = V;
    
    dx = width/nc;
    dy = height/nr;
    
    hex = ((hbar/pow(dx,2))*(hbar/(efm*mass)))/exrg;
    hey = ((hbar/pow(dy,2))*(hbar/(efm*mass)))/exrg;
}

GreensFunction::GreensFunction(void) {
    nc = 0;
    nr = 0;
}


// 2D Lead Green's function
cx_fmat GreensFunction::leadGreensFunction(float E) {
    float                   Vx, Vy, dL;
    complex<float>          p,q;
    const complex<float>    img = complex<float>(0,1);
    const float             pi = 3.14159265358979323;
    
    cx_fmat G, U;
    
    G = zeros<cx_fmat>(nr,nc);
    U = zeros<cx_fmat>(nr,nc);
    
    dL = 1.0/(Potential.n_rows+1);
    Vx = -hex;
    Vy = -hey;
        
    q = 2*abs(Vx);
    for(int i = 1; i <= nr; i ++) {
        p = E + complex<float>(2,0)*(Vx+Vy) + complex<float>(2,0)*abs(Vy)*cos(pi*i*dL) + eta*img;
        G(i-1,i-1) = (complex<float>(2,0)*p/pow(q,2))*(complex<float>(1,0)-sqrt(complex<float>(1,0)-q*q/(p*p)));
    }
    
    for(int j = 0; j < nc; j ++) 
        for(int i = 0; i < nr; i ++) 
            U(i,j) = sqrt(2.0*dL)*sin((j*pi*dL)*i);
    
    return U*G*U.t();
}

// Hamiltonian
cx_fmat GreensFunction::Hamiltonian(int n) {
    cx_fmat H;
    
    H = zeros<cx_fmat>(Potential.n_rows, Potential.n_rows);
    H.diag(1) = -hey*ones<cx_fmat>(nr-1);
    H.diag(-1) = -hey*ones<cx_fmat>(nr-1);
    H.diag(0) = 2*(hex+hey) + conv_to<cx_fmat>::from(Potential.col(n));
    
    return H;
}

// General transmission function
float GreensFunction::Transmission(float Energy, float B) {
    cx_fmat V, G_lead, G_nn, G_Ln, Sigma, P, Gama, G;
    
    cx_fmat I = eye<cx_fmat>(nr,nr);
    complex<float> img = complex<float>(0,1);
    complex<float> VB;
    
    // Coupling matrix
    VB = img*exrg*B*dx*dy/hbar;
    V = eye<cx_fmat>(nr,nr);
    for(int n = 0; n < nr; n ++)
        V(n,n) = -hex*exp(complex<float>(n+1,0)*VB);
    
    // Begin with left lead
    G_lead = leadGreensFunction(Energy);
    G_nn = G_lead;
    G_Ln = G_lead;
    
    /* Loop over columns
     * Sigma = V+.G_nn.V
     * G_nn = [(E+i.eta).I - H - S]^(-1)
     * G_Ln = G_Ln.V.G_nn */
    for(int n = 0; n < nc; n ++) {
        Sigma = conj(V)*G_nn*V;
        P = (Energy+img*eta)*I-Hamiltonian(n) - Sigma;
        G_nn = inv(P);
        G_Ln = G_Ln*V*G_nn;
    }
    
    /* Add right lead
     * [G_lead.V.G_nn.V).G_nn = G_lead
     * G = G_Ln.V.G_nn */
    Sigma = conj(V)*G_nn*V;
    P = I - G_lead*Sigma;
    G_nn = solve(P,G_lead);
    G = G_Ln*V*G_nn;
    
    /* Compute Gamma matrix
     * S = V+.G_lead.V
     * Gamma = i.(S-S+) */
    Sigma = conj(V)*G_lead*V;
    Gama = img*(Sigma - conj(Sigma));
    
    // r = trace(Gama.G.Gama.G+)
    P = Gama*G*Gama*conj(G);
    return trace(real(P));
}

// Energy Sweep
vec GreensFunction::EnergySweep(float Emin, float Emax, int N,float B) {
    vec     T(N);
    float   Energy;
    
    for(int i = 0; i < N; i ++) {
        Energy = Emin + (Emax - Emin)*i/N;
        T(i) = Transmission(Energy,B);
    }
    
    return T;
}
        
// Density of States
mat GreensFunction::DOS(float Energy, float B) {
    cx_fmat     V, Gr, Gnn, Sigma, Gama, G, P;
    cx_fcube    M;
    mat         D;
    
    cx_fmat I = eye<cx_fmat>(nr,nr);
    complex<float> img = complex<float>(0,1);
    complex<float> VB;    
    
    // Initializations
    D = zeros(nr,nc);
    M = zeros<cx_fcube>(nr,nr,nc+1);
    
    // Coupling matrix
    VB = img*exrg*B*dx*dy/hbar;
    V = eye<cx_fmat>(nr,nr);
    for(int n = 0; n < nr; n ++)
        V(n,n) = -hex*exp(complex<float>(n+1,0)*VB);    
    
    // Lead Green's function
    Gr = leadGreensFunction(Energy);
    M.slice(nc) = Gr;
    
    // Backward Propagation
    Gnn = Gr;
    for(int j = nc-1; j >= 0; j --) {
        Sigma = V*Gnn*conj(V);
        P = (Energy+img*eta)*I - Hamiltonian(j) - Sigma;
        Gnn = inv(P);
        M.slice(j) = Gnn;
    }
    
    // Forward + Merging
    for(int j = 0; j < nc; j ++) {
        // Forward
        Sigma = conj(V)*Gnn*V;
        P = (Energy+img*eta)*I - Hamiltonian(j) - Sigma;
        Gnn = inv(P);
        
        // Merging
        Sigma = V*M.slice(j+1)*conj(V);
        P = I - Gnn*Sigma;
        G = solve(P,Gnn);
        
        // DOS
        D.col(j) = -conv_to<mat>::from(diagvec(imag(G)));
    }
    
    return D;
}
    
    

File: Julia/Green.jl
using LinearAlgebra
using Plots
#using PyPlot

mutable struct GreenParameters
    nrows   :: Integer
    ncols   :: Integer
    dx      :: Real
    dy      :: Real
    hex     :: Real
    hey     :: Real
    Pot     :: Matrix{Float64}
    iη      :: ComplexF64
end


#-----------------------------#
# Initialize Green's function #
#-----------------------------#
function Green(Pot,height,width,effective_mass)
    ħ = 1.054571817E-34     # J.s
    q = 1.602176634E-19     # C
    mo = 9.1093837015E-31   # kg
    
    rows,cols = size(Pot)
    
    dx = width/cols
    dy = height/rows
    
    hex = ((ħ/dx^2)*(ħ/(effective_mass*mo)))/q
    hey = ((ħ/dy^2)*(ħ/(effective_mass*mo)))/q
    
    G = GreenParameters(rows,cols,dx,dy,hex,hey,Pot,1E-9*im)
    
    #println("Hopping energy x: ",1E3*hex," [meV]")
    #println("Hopping energy y: ",1E3*hey," [meV]")
    
    return G
end 

#-----------------------#
# Lead Green's function #
#-----------------------#
function leadGF(E,par::GreenParameters)
    π = 3.14159265358979323
    
    G = zeros(ComplexF64,par.nrows,par.nrows)
    U = zeros(ComplexF64,par.nrows,par.nrows)
    
    dL = 1.0/(par.nrows+1)
    Vx = -par.hex
    Vy = -par.hey
    q = 2*abs(Vx)
            
    for i in 1:par.nrows
        p = E + 2*(Vx+Vy) + 2*abs(Vy)*cos(Complex(π*i*dL)) + par.iη;
        G[i,i] = (2*p/q^2)*(1.0 - sqrt(Complex(1.0 - (q/p)^2)));
        
        for j in 1:par.nrows
            U[i,j] = sqrt(Complex(2*dL))*sin((j*pi*dL)*i);
        end
    end
        
    return U*G*U';    
end

#-------------#
# Hamiltonian #
#-------------#
function Hamiltonian(n,par::GreenParameters)
    P = par.Pot[:,n]
    v = -par.hey*ones(par.nrows - 1)
    H = Tridiagonal(v, 2*(par.hex + par.hey)*ones(par.nrows) + P, v)
    
    return H
end

#-------------------#
# Density of States #
#-------------------#
function DOS(Energy, B, par::GreenParameters)
    h = 6.62607015E-34      # J.s
    q = 1.602176634E-19     # C

    M = zeros(ComplexF64,par.nrows,par.nrows,par.ncols+1)
    D = zeros(par.nrows, par.ncols)
    
    Φ = B*par.dx*par.dy
    Φo = h/q
    τ = -I(par.nrows)*par.hex*exp(im*Φ/Φo)
    τl = τ'
    
    Gr = leadGF(Energy, par)
    M[:,:,par.ncols+1] = Gr
        
    # Backward propagation
    Gnn = Gr
    for j in par.ncols:-1:1
        Σ = τ*Gnn*τl
        H = Hamiltonian(j,par)
        Gnn = inv((Energy + par.iη)*I(par.nrows) - H - Σ)
        M[:,:,j] = Gnn
    end
   
    # Forward progpagation
    Gnn = Gr
    for j in 1:par.ncols
        # propagate
        Σ = τl*Gnn*τ
        H = Hamiltonian(j,par)
        Gnn = inv((Energy + par.iη)*I(par.nrows) - H - Σ)
        
        #merge
        Σ = τ*M[:,:,j+1]*τl
        G = (I(par.nrows) - Gnn*Σ) \ Gnn
        
        #DOS
        D[:,j] = -imag(diag(G))
    end
    
    return D
end         

#--------------#
# Transmission #
#--------------#
function Transmission(Energy, B, par::GreenParameters)
    h = 6.62607015E-34      # J.s
    q = 1.602176634E-19     # C
    
    Φ = B*par.dx*par.dy
    Φo = h/q
    τ = -I(par.nrows)*par.hex*exp(im*Φ/Φo)
    τl = τ'    
        
    # Begin with left lead
    G_lead = leadGF(Energy, par)
    G_nn = G_lead
    G_Ln = G_lead
    
    # Loop over columns
    for n in 1:par.ncols
        Σ = τl*G_nn*τ
        H = Hamiltonian(n,par)
        G_nn = inv((Energy + par.iη)*I(par.nrows) - H - Σ)
        G_Ln = G_Ln*τ*G_nn
    end
    
    # Add right lead
    Σ = τl*G_nn*τ
    G_nn = (I(par.nrows)-G_lead*Σ) \ G_lead
    G = G_Ln*τ*G_nn
    
    # Compute Γ matrix
    Σ = τl*G_lead*τ
    Γ = im*(Σ - Σ')
    T = Γ*G*Γ*G'
    t = tr(real(T))
    
    return t
end


#==============================================================#
#                           MAIN                               #
#==============================================================#
L = 100
Pot = ones(L,L)
l2 = L÷2 

for j in 1:L
    for i in 1:L
        r = ((i-l2)/l2)^2 + ((j-l2)/l2)^2
        if (r <= 1 || (i > 80 && i < 120))
            Pot[i,j] = 0
        end
    end
end

pars = Green(Pot,675E-9,675E-9,0.041)

t = []
E = []
B = 1
anim = @animate for i in 1:200
#for i in 1:200
    Energy = 50E-4 + i*1E-5

    #push!(t,Transmission(Energy,pars))
    #push!(E,Energy)
    
    D = DOS(Energy,B,pars)
    heatmap(D,c=:redsblues)
end

#plot(E,t)
#show()

gif(anim, "test.gif", fps = 5)
